applications/IgaApplication/tests/shell_kl_discrete_element_test.py

```diff
@@ -0,0 +1,294 @@
+            shapes.Compute(surface.KnotsU(), surface.KnotsV(), u, v)
+
+            node_indices = [surface.Node(u, v).Id
+                for u, v in shapes.NonzeroPoleIndices]
+
+            element = model_part.CreateNewElement('ShellKLDiscreteElement',
+                i + 1, node_indices, shell_properties)
+
+            n_0 = Vector(shapes.NumberOfNonzeroPoles)
+            n_1 = Matrix(shapes.NumberOfNonzeroPoles, 2)
+            n_2 = Matrix(shapes.NumberOfNonzeroPoles, 3)
+
+            for j, (poleU, poleV) in enumerate(shapes.NonzeroPoleIndices):
+                indexU = poleU - shapes.FirstNonzeroPoleU
+                indexV = poleV - shapes.FirstNonzeroPoleV
+                n_0[j   ] = shapes(0, indexU, indexV)
+                n_1[j, 0] = shapes(1, indexU, indexV)
+                n_1[j, 1] = shapes(2, indexU, indexV)
+                n_2[j, 0] = shapes(3, indexU, indexV)
+                n_2[j, 1] = shapes(5, indexU, indexV)
+                n_2[j, 2] = shapes(4, indexU, indexV)
+
+            element.SetValue(INTEGRATION_WEIGHT, weight)
+            element.SetValue(SHAPE_FUNCTION_VALUES, n_0)
+            element.SetValue(SHAPE_FUNCTION_LOCAL_DERIVATIVES, n_1)
+            element.SetValue(SHAPE_FUNCTION_LOCAL_SECOND_DERIVATIVES, n_2)
+
+        # add dofs
+
+        VariableUtils().AddDof(DISPLACEMENT_X, REACTION_X, model_part)
+        VariableUtils().AddDof(DISPLACEMENT_Y, REACTION_Y, model_part)
+        VariableUtils().AddDof(DISPLACEMENT_Z, REACTION_Z, model_part)
+
+        # apply dirichlet conditions
+
+        surface.Node(0, 0).Fix(DISPLACEMENT_X)
+        surface.Node(0, 0).Fix(DISPLACEMENT_Y)
+        surface.Node(0, 0).Fix(DISPLACEMENT_Z)
+
+        surface.Node(0, 1).Fix(DISPLACEMENT_X)
+        surface.Node(0, 1).Fix(DISPLACEMENT_Y)
+        surface.Node(0, 1).Fix(DISPLACEMENT_Z)
+
+        surface.Node(0, 2).Fix(DISPLACEMENT_X)
+        surface.Node(0, 2).Fix(DISPLACEMENT_Y)
+        surface.Node(0, 2).Fix(DISPLACEMENT_Z)
+
+        surface.Node(1, 0).Fix(DISPLACEMENT_Z)
+
+        surface.Node(1, 1).Fix(DISPLACEMENT_Z)
+
+        surface.Node(1, 2).Fix(DISPLACEMENT_Z)
+
+        # apply neumann conditions
+
+        prop = Properties(2)
+        model_part.AddProperties(prop)
+
+        node_n0 = surface.Node(surface.NbPolesU - 1, 0)
+        node_nm = surface.Node(surface.NbPolesU - 1, surface.NbPolesV - 1)
+
+        model_part.CreateNewCondition('PointLoadCondition3D1N', 1, [node_n0.Id],
+            prop)
+        node_n0.SetSolutionStepValue(POINT_LOAD_Z, -50)
+
+        model_part.CreateNewCondition('PointLoadCondition3D1N', 2, [node_nm.Id],
+            prop)
+        node_nm.SetSolutionStepValue(POINT_LOAD_Z, -50)
+
+        # setup solver
+
+        model_part.SetBufferSize(1)
+
+        time_scheme = ResidualBasedIncrementalUpdateStaticScheme()
+
+        linear_solver = new_linear_solver_factory.ConstructSolver(Parameters(
+            r'{"solver_type": "eigen_sparse_lu"}'))
```

Comment:
This cannot be used in a test, please use another one
=> test will fail if one does not have Eigen

Reply:
I was not sure which solver I should use since the `ExternalSolverApplication` is also a separate application. Is there a direct solver in the core?

Reply:
You can use the one in amgcl. It's wrapped in kratos/linear_solvers/skyline_lu_custom_scalar_solver.h